pkg/kubelet/dockertools/manager.go

```diff
@@ -1748,16 +1772,69 @@ func (dm *DockerManager) clearReasonCache(pod *api.Pod, container *api.Container
 	dm.reasonCache.Remove(pod.UID, container.Name)
 }
 
+// TODO (random-liu) Move this to container/runtime.go
+
+// SyncAction indicates different kind of actions in SyncPod(). Now there are only actions
+// about start/kill container and setup/teardown network.
+type SyncAction int
+
+const (
+	StartContainer SyncAction = iota
+	KillContainer
+	SetupNetwork
+	TeardownNetwork
+	// Add more later if needed
+)
+
+// SyncResult is the result of sync action.
+type SyncResult struct {
+	// The associated action of the result
+	Action SyncAction
+	// The target of the action, now the target can only be:
+	//  * Container: Target should be container name
+	//  * Network: Target is useless now, we just set it as pod full name now
+	Target interface{}
+	// Brief error reason
+	Error error
+	// Human readable error reason
+	Message string
+}
+
+func (r *SyncResult) Fail(err error, msg string) {
+	r.Error, r.Message = err, msg
+}
+
+// Now the PodSyncResult is just a slice holding all the SyncResult
+type PodSyncResult []*SyncResult
+
+func (p PodSyncResult) NewSyncResult(action SyncAction, target interface{}) *SyncResult {
+	result := &SyncResult{Action: action, Target: target}
+	p = append(p, result)
+	return result
+}
+
+func (p PodSyncResult) AddSyncResult(result ...*SyncResult) {
+	p = append(p, result...)
+}
+
 // Sync the running pod to match the specified desired pod.
-func (dm *DockerManager) SyncPod(pod *api.Pod, _ kubecontainer.Pod, _ api.PodStatus, podStatus *kubecontainer.PodStatus, pullSecrets []api.Secret, backOff *util.Backoff) error {
+func (dm *DockerManager) SyncPod(pod *api.Pod, runningPod kubecontainer.Pod, apiPodStatus api.PodStatus, podStatus *kubecontainer.PodStatus, pullSecrets []api.Secret, backOff *util.Backoff) error {
+	err, _ := dm.syncPodWithSyncResult(pod, runningPod, apiPodStatus, podStatus, pullSecrets, backOff)
+	return err
+}
+
+// (random-liu) This is just a temporary function, will be removed when we acturally add PodSyncEvent
+func (dm *DockerManager) syncPodWithSyncResult(pod *api.Pod, _ kubecontainer.Pod, _ api.PodStatus, podStatus *kubecontainer.PodStatus, pullSecrets []api.Secret, backOff *util.Backoff) (error, PodSyncResult) {
 	start := time.Now()
 	defer func() {
 		metrics.ContainerManagerLatency.WithLabelValues("SyncPod").Observe(metrics.SinceInMicroseconds(start))
 	}()
 
+	var result PodSyncResult
+
 	containerChanges, err := dm.computePodContainerChanges(pod, podStatus)
 	if err != nil {
-		return err
+		return err, result
```

Comment:
question: what sync result is expected when computePodContainerChanges fails here?